### PR TITLE
AES GCM: implement streaming

### PR DESCRIPTION
**Pull request by SparkiDev**
Updated EVP layer too.
Assembly for x64 updated to include streaming.

### CODE REVIEW

File: wolfcrypt/src/aes.c [hunk 1]
```diff
@@ -7379,6 +7707,1004 @@ int wc_AesGcmDecrypt(Aes* aes, byte* out, const byte* in, word32 sz,
...
+    /* Set the key if passed in. */
+    if ((ret == 0) && (key != NULL)) {
+        ret = wc_AesGcmSetKey(aes, key, len);
+    }
+
+    if (ret == 0) {
+        /* Setup with IV if needed. */
+        if (iv != NULL) {
+            /* Cache the IV in AES GCM object. */
+            XMEMCPY((byte*)aes->reg, iv, ivSz);
```
Comment (guidovranken) [inline wolfcrypt/src/aes.c:8487]: Check if ivSz is valid (e.g. `<= sizeof(aes->reg)`) to prevent OOB write.

Comment (SparkiDev) [reply wolfcrypt/src/aes.c:8487]: Push a fix for this. Checking IV size is not greater than AES_BLOCK_SIZE.

**Comment by SparkiDev** [comment]
The API doesn't support mixing AAD and cipher text.
Always put AAD first and then the ciphertext.
Will push an update to the comments.

File: wolfcrypt/src/aes.c (hunk 1 above)
Comment (guidovranken) [reply wolfcrypt/src/aes.c:8487]: Tested and confirmed fixed.

Please also reject IV of size 0, as the spec prohibits this.

**Comment by guidovranken** [comment]
> The API doesn't support mixing AAD and cipher text.
> Always put AAD first and then the ciphertext.
> Will push an update to the comments.

Thanks, fixed this in my fuzzer and the problem is gone.

However, I get the following discrepancy when comparing against Botan:

```
operation name: SymmetricEncrypt
cleartext: {}
aad: {0x6d, 0x7e, 0xf7, 0xf3, 0xd2} (5 bytes)
cipher iv: {0xa3, 0xb6, 0x3f, 0xd9, 0x41, 0xa2, 0x0c, 0xc8, 0xbc, 0xbe, 0x0e, 0x9c, 0x79, 0x32, 0x27, 0xf8} (16 bytes)
cipher key: {0x3d, 0x87, 0x61, 0xb3, 0x37, 0xe6, 0x54, 0x0f, 0xfc, 0xc5, 0x26, 0x57, 0x71, 0xc4, 0x39, 0xb6, 
 0x7d, 0x89, 0xf5, 0x4f, 0x06, 0xc0, 0xc2, 0x47, 0xee, 0x07, 0x04, 0xc9, 0x22, 0x0b, 0x1e, 0xf7} (32 bytes)
cipher: AES_256_GCM
ciphertextSize: 6
tagSize: 8

Module wolfCrypt result:

ciphertext = {}
tag = {0x40, 0x7c, 0x5f, 0x1d, 0x80, 0x84, 0x61, 0xe2} (8 bytes)

Module Botan result:

ciphertext = {}
tag = {0xbc, 0xd7, 0x55, 0x36, 0x62, 0x05, 0x50, 0x26} (8 bytes)
```

Reproducer:

```c
#include <wolfssl/options.h>
#include <wolfssl/wolfcrypt/aes.h>

#define CF_CHECK_EQ(expr, res) if ( (expr) != (res) ) { goto end; }

int main(void)
{
    Aes ctx;
    const unsigned char in[] = {};
    unsigned char out[1];
    const unsigned char aad[] = {0x6d, 0x7e, 0xf7, 0xf3, 0xd2};
    const unsigned char key[32] = {0x3d, 0x87, 0x61, 0xb3, 0x37, 0xe6, 0x54, 0x0f, 0xfc, 0xc5, 0x26, 0x57, 0x71, 0xc4, 0x39, 0xb6,
                                   0x7d, 0x89, 0xf5, 0x4f, 0x06, 0xc0, 0xc2, 0x47, 0xee, 0x07, 0x04, 0xc9, 0x22, 0x0b, 0x1e, 0xf7};
    const unsigned char iv[16] = {0xa3, 0xb6, 0x3f, 0xd9, 0x41, 0xa2, 0x0c, 0xc8, 0xbc, 0xbe, 0x0e, 0x9c, 0x79, 0x32, 0x27, 0xf8};
    unsigned char tag[8];
    CF_CHECK_EQ(wc_AesGcmInit(&ctx, key, sizeof(key), iv, sizeof(iv)), 0);
    /* Pass AAD */
    CF_CHECK_EQ(wc_AesGcmEncryptUpdate(&ctx, NULL, NULL, 0, aad, sizeof(aad)), 0);
    /* Pass cleartext */
    CF_CHECK_EQ(wc_AesGcmEncryptUpdate(&ctx, out, in, sizeof(in), NULL, 0), 0);
    CF_CHECK_EQ(wc_AesGcmEncryptFinal(&ctx, tag, sizeof(tag)), 0);
    for (size_t i = 0; i < sizeof(tag); i++) {
        printf("%02X ", tag[i]);
    }
    printf("\n");

end:
    return 0;
}
```

**Comment by SparkiDev** [comment]
Hi @guidovranken 

I've fixed this use case of no cipher text and AAD not a multiple of a block size.

Thanks!

**Comment by guidovranken** [comment]
Great. Did you also think of the empty IV? (I haven't tested your latest commit yet) https://github.com/wolfSSL/wolfssl/pull/3841#discussion_r589287337

**Comment by SparkiDev** [comment]
Hi @guidovranken,

An empty IV means re-use the IV setup before or you will be making another call with the IV.
You can either set the key and/or the IV in the API.
I'll add a check to ensure an IV has been set in the update functions.

**Comment by SparkiDev** [comment]
Added checking of key and IV/nonce being set and a set of tests in api.c.

File: wolfcrypt/src/aes.c (hunk 1 above)
Comment (SparkiDev) [reply wolfcrypt/src/aes.c:8520]: Done!

**Comment by toddouska** [comment]
@dgarske please review and then assign to me after approval.

**Comment by SparkiDev** [comment]
Rebased to master to fix conflicts.

**Review by dgarske — CHANGES_REQUESTED** [review]
Just a few minor items, otherwise is in great shape.

File: configure.ac [hunk 2]
```diff
@@ -1295,6 +1295,11 @@ AC_ARG_ENABLE([aesgcm],
     [ ENABLED_AESGCM=$enableval ],
     [ ENABLED_AESGCM=yes ]
     )
+AC_ARG_ENABLE([aesgcm-stream],
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply configure.ac:1298]: I tried but FIPS stuffs things up.
It replaces the file, sometimes.
Any thoughts on how to get it to work?

File: wolfssl/wolfcrypt/aes.h [hunk 3]
```diff
@@ -361,6 +376,24 @@ WOLFSSL_API int wc_AesEcbDecrypt(Aes* aes, byte* out,
                                    const byte* iv, word32 ivSz,
                                    const byte* authTag, word32 authTagSz,
                                    const byte* authIn, word32 authInSz);
+#ifdef WOLFSSL_AESGCM_STREAM
+WOLFSSL_API int wc_AesGcmInit(Aes* aes, const byte* key, word32 len,
+        const byte* iv, word32 ivSz);
+
+#define wc_AesGcmEncryptInit    wc_AesGcmInit
+WOLFSSL_API int wc_AesGcmEncryptInit_ex(Aes* aes, const byte* key, word32 len,
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply wolfssl/wolfcrypt/aes.h:384]: For FIPS, the IV is generated and then copied out for decrypt.
Decrypt doesn't generate IV.
(See wc_AesGcmEncrypt_ex())

File: wolfcrypt/src/aes.c [hunk 4]
```diff
@@ -8210,6 +9659,15 @@ int wc_AesInit(Aes* aes, void* heap, int devId)
     aes->aadLen = 0;
 #endif
 #endif
+
+#ifdef WOLFSSL_AESGCM_STREAM
+    aes->keylen = 0;
+    aes->nonceSz = 0;
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply wolfcrypt/src/aes.c:9665]: Fixed configure.ac

File: wolfssl/wolfcrypt/aes.h [hunk 5]
```diff
@@ -361,6 +376,24 @@ WOLFSSL_API int wc_AesEcbDecrypt(Aes* aes, byte* out,
                                    const byte* iv, word32 ivSz,
                                    const byte* authTag, word32 authTagSz,
                                    const byte* authIn, word32 authInSz);
+#ifdef WOLFSSL_AESGCM_STREAM
+WOLFSSL_API int wc_AesGcmInit(Aes* aes, const byte* key, word32 len,
+        const byte* iv, word32 ivSz);
+
+#define wc_AesGcmEncryptInit    wc_AesGcmInit
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply wolfssl/wolfcrypt/aes.h:383]: Done.

File: wolfcrypt/src/evp.c [hunk 6]
```diff
@@ -5275,6 +5378,50 @@ int wolfSSL_EVP_MD_type(const WOLFSSL_EVP_MD *md)
...
+                            len, NULL, 0);
+                        if (ret == 0) {
+                            ret = wc_AesGcmDecryptFinal(&ctx->cipher.aes,
+                                ctx->authTag, ctx->authTagSz);
+                        }
+                    }
+                }
+                if (ret == 0) {
+                    if (ctx->enc) {
+                        /* Calcuate authentication tag. */
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply wolfcrypt/src/evp.c:5414]: fixed

File: wolfcrypt/src/evp.c [hunk 7]
```diff
@@ -5124,8 +5221,13 @@ int wolfSSL_EVP_MD_type(const WOLFSSL_EVP_MD *md)
                                              int ivLen)
     {
         WOLFSSL_ENTER("wolfSSL_EVP_CIPHER_CTX_set_iv_length");
-        if (ctx)
+        if (ctx) {
             ctx->ivSz= ivLen;
+        #ifdef WOLFSSL_AESGCM_STREAM
+            /* Store IV len in AES object for init. */
+            ctx->cipher.aes.nonceSz = ivLen;
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply wolfcrypt/src/evp.c:5228]: Don't appear to actually need this code.
Removed.

File: wolfcrypt/src/evp.c [hunk 8]
```diff
@@ -4423,14 +4509,25 @@ int wolfSSL_EVP_MD_type(const WOLFSSL_EVP_MD *md)
...
                 return WOLFSSL_FAILURE;
             }
+#endif /* !WOLFSSL_AESGCM_STREAM */
             if (iv && wc_AesGcmSetExtIV(&ctx->cipher.aes, iv, GCM_NONCE_MID_SZ)) {
                 WOLFSSL_MSG("wc_AesGcmSetExtIV() failed");
                 return WOLFSSL_FAILURE;
             }
+#ifdef WOLFSSL_AESGCM_STREAM
+            /* Initialize with key and IV if available. */
+            if (wc_AesGcmInit(&ctx->cipher.aes, key,
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply wolfcrypt/src/evp.c:4524]: Done as requested.

File: wolfssl/wolfcrypt/aes.h [hunk 9]
```diff
@@ -250,6 +250,21 @@ struct Aes {
     silabs_aes_t ctx;
 #endif
     void*  heap; /* memory hint to use */
+#ifdef WOLFSSL_AESGCM_STREAM
+    ALIGN32 byte initCtr[AES_BLOCK_SIZE];
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply wolfssl/wolfcrypt/aes.h:254]: OK. Tell me what you think of the change.
Only needs to be 16 byte aligned as using xmm registers. Not allocating when using AESNI.
One buffer for all data and macros for the fields.

**Review by dgarske — APPROVED** [review]
Great work. Over to Todd for final.

File: wolfssl/wolfcrypt/aes.h [hunk 10]
```diff
@@ -250,6 +250,21 @@ struct Aes {
     silabs_aes_t ctx;
 #endif
     void*  heap; /* memory hint to use */
+#ifdef WOLFSSL_AESGCM_STREAM
+#if !defined(WOLFSSL_SMALL_STACK) || defined(WOLFSSL_AESNI)
+    ALIGN16 byte streamData[5 * AES_BLOCK_SIZE];
+#else
+    byte*        streamData;
```
Comment (toddouska) [inline wolfssl/wolfcrypt/aes.h:257]: Is this free'd during Final in SMALL_STACK?  I don't see where.

Comment (SparkiDev) [reply wolfssl/wolfcrypt/aes.h:257]: No freed in wc_AesFree().
Didn't want to allocate and free this every time.
Just saving the memory from being allocated for non AES-GCM streaming operations.
wc_AesInit() must be called.

**Review by toddouska — APPROVED** [review]

**Pull request merged**